provider/pkg/gen/schema_test.go

```diff
@@ -0,0 +1,189 @@
+package gen
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/jtacoma/uritemplates"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/pulumi/pulumi/pkg/v3/codegen"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi-google-native/provider/pkg/resources"
+
+	"github.com/stretchr/testify/require"
+)
+
+var root = path.Join("..", "..", "..")
+
+// This is a validation test to make sure that the metadata generated at schema generation stands-up
+// to the assumptions made for operation URL resolution at runtime in the provider.
+// NOTE this will need to be kept up-to-date with the provider.go implementation but this framework gives
+// valuable confidence in correctness around metadata generation without destabalizing the provider at runtime.
+func TestMetadata_Operations(t *testing.T) {
+	cwd, err := os.Getwd()
+	require.NoError(t, err)
+	t.Logf("Current directory: %s", cwd)
+	var fileNames []string
+	require.NoError(t, filepath.Walk(filepath.Join(root, "discovery"), func(path string, info os.FileInfo,
+		err error) error {
+		if err != nil {
+			return err
+		}
+		if !info.IsDir() {
+			fileNames = append(fileNames, path)
+		}
+		return nil
+	}))
+
+	metadata, err := loadMetadata()
+	require.NoError(t, err)
+
+	pkgSpec, err := loadSchema()
+	require.NoError(t, err)
+
+	for _, fileName := range fileNames {
+		document, err := readDiscoveryDocument(fileName)
+		require.NoError(t, err)
+
+		if document.Name == "" {
+			continue
+		}
+		module := fmt.Sprintf("%s/%s", document.Name, document.Version)
+		res, ops, err := findResources(fileName, document.Resources, document.Schemas)
+		require.NoError(t, err)
+
+		resourcesForModule := codegen.NewStringSet()
+		for tok := range metadata.Resources {
+			if strings.Contains(tok, module) {
+				resourcesForModule.Add(tok)
+			}
+		}
+
+		for _, tok := range resourcesForModule.SortedValues() {
+			mdResource := metadata.Resources[tok]
+			resource := pkgSpec.Resources[tok]
+			validateOpTemplate := func(resourceOperationType string, op *resources.Operations,
+				ddResource *discoveryDocumentResource) {
+				if op == nil || op.Template == "" {
+					return
+				}
+				tmpl, err := uritemplates.Parse(op.Template)
+				require.NoError(t, err)
+				paramNames := codegen.NewStringSet(tmpl.Names()...)
+
+				var referencedOp *operation
+				switch resourceOperationType {
+				case "CREATE":
+					if ddResource.createMethod.Response.Ref != "" {
+						if createOp, ok := ops[ddResource.createMethod.Response.Ref]; ok {
+							referencedOp = createOp
+						}
+					}
```

Comment:
Do we also need to validate READ?

Reply:
Reads don't result in delayed/long running operations like the other verbs.